Upstream-Status: Backport

* Backported from the ToT: https://crrev.com/c/1407011

Signed-off-by: Maksim Sisov <msisov@igalia.com>
---
From d2dca30aeac3ab753a3e04a77cc26446b89d9494 Mon Sep 17 00:00:00 2001
From: Nick Diego Yamane <nickdiego@igalia.com>
Date: Mon, 14 Jan 2019 15:10:52 +0000
Subject: [PATCH 06/16] [ozone/wayland] Fix event routing issue when multiple
 windows are open

Currently ozone wayland implementation uses singletons for bridging
wayland input objects (e.g: WaylandPointer, WaylandKeyboard, WaylandTouch),
each of them inject events into chromium event system. Additionally
they are currently responsible for controlling implicit_grab flag stored
on WaylandWindow instances.
This particular issue happens when new windows are created and the user
tries to trigger any action in browser UI (e.g: create new tab) of the
newly created window, what is currently leading to actions to be taken in
the wrong window (e.g: the new tab is open in the wrong window).
Under the hood, the bug happens because the implicit_grab flag
is not correctly unset for the window that lost focus. So the fix
consists on properly unsetting it on WaylandPointer.

Bug: 921009
Change-Id: I25092fa211849f0345895b88681eaded52f09792
Reviewed-on: https://chromium-review.googlesource.com/c/1407011
Reviewed-by: Maksim Sisov <msisov@igalia.com>
Commit-Queue: Nick Yamane <nickdiego@igalia.com>
Cr-Commit-Position: refs/heads/master@{#622459}
---
 ui/ozone/platform/wayland/wayland_pointer.cc | 1 +
 1 file changed, 1 insertion(+)

diff --git a/ui/ozone/platform/wayland/wayland_pointer.cc b/ui/ozone/platform/wayland/wayland_pointer.cc
index f977f636681a..4e7940cd8e95 100644
--- a/ui/ozone/platform/wayland/wayland_pointer.cc
+++ b/ui/ozone/platform/wayland/wayland_pointer.cc
@@ -80,6 +80,7 @@ void WaylandPointer::Leave(void* data,
   if (surface) {
     WaylandWindow* window = WaylandWindow::FromSurface(surface);
     window->set_pointer_focus(false);
+    window->set_has_implicit_grab(false);
     pointer->window_with_pointer_focus_ = nullptr;
   }
 }
-- 
2.17.1

